L1Trigger/L1THGCalUtilities/python/hgcalTriggerChains.py

```diff
@@ -42,15 +48,42 @@ def register_chain(self, vfe, concentrator, backend1, backend2, selector='', ntu
         if ntuple!='' and not ntuple in self.ntuple: 
             raise KeyError('{} not registered as ntuplizer'.format(ntuple))
         self.chain.append( (vfe, concentrator, backend1, backend2, selector, ntuple) )
-    
+
+    def register_truth_chain(self, vfe, truth_prod, selector='', ntuple=''):
+        if not vfe in self.vfe: 
+            raise KeyError('{} not registered as VFE producer'.format(vfe))
+        if not truth_prod in self.truth_prod: 
+            raise KeyError('{} not registered as truth producer'.format(truth_prod))
+        self.truth_chain.append( (vfe, truth_prod, selector, ntuple) )
+
+    def register_truth_be_chain(self, vfe, truth_prod, backend1, backend2, selector='', ntuple=''): 
+        if not vfe in self.vfe: 
+            raise KeyError('{} not registered as VFE producer'.format(vfe))
+        if not truth_prod in self.truth_prod: 
+            raise KeyError('{} not registered as truth producer'.format(truth_prod))
+        if not backend1 in self.backend1: 
+            raise KeyError('{} not registered as backend1 producer'.format(backend1))
+        if not backend2 in self.backend2: 
+            raise KeyError('{} not registered as backend2 producer'.format(backend2))
+        if selector!='' and not selector in self.selector:
+            raise KeyError('{} not registered as selector'.format(selector))
+        if ntuple!='' and not ntuple in self.ntuple: 
+            raise KeyError('{} not registered as ntuplizer'.format(ntuple))
+        self.truth_be_chain.append( (vfe, truth_prod, backend1, backend2, selector, ntuple) )
 
     def create_sequences(self, process):
+        from L1Trigger.L1THGCalUtilities.caloTruthCellsProducer_cfi import caloTruthCellsProducer
```

Comment:
- All this should only be done if truth chains are requested, and if the truth cells producer is not already in the process, similarly to what is done below:
``` python
        if not hasattr(process, 'hgcalTriggerSelector'):
            process.load('L1Trigger.L1THGCalUtilities.HGC3DClusterGenMatchSelector_cff')
```
- And I think it will create problems if both chains with and without truth are used, because the standard `hgcalTriggerPrimitives` sequence is modified. So we'll probably need to prevent this and add an exception if standard chains and truth chains are called in the same process (or find a way to make them work together)

Reply:
I agree. What I just did was to define an hgcalTriggerPrimitivesTruth and to add that as a separate step. In the user configuration, you'd need: 

process.hgcl1tpg_truth_step = cms.Path(process.hgcalTriggerPrimitivesTruth)
process.schedule = cms.Schedule(process.hgcl1tpg_step, process.hgcl1tpg_truth_step, process.selector_step, process.ntuple_step)

Does this seem like a reasonable solution?

Reply:
Not really because the truth step should not be added for standard sequences.

Reply:
I separated the standard sequence from truth sequence. After thinking about it more, I think the "full truth" and "truth_be" cases can be handled together. We would just need to have a dummy option for the 3D clustering. So I'd propose adding that in.

Reply:
I think it's a good thing to separate clearly the two sequences.

If the "full truth" and the "truth be" are merged, given how it is currently implemented, the clusters used in the ntuples are the `backend2` clusters. Which means that the sim clusters are produced but there is no way to connect them to the ntuples anymore.